### PR TITLE
Sync all user data to local db table

### PR DESCRIPTION
**Pull request by henrikskog**
**Changes approach for how we store user data. Takes inspiration from the auth0 documentation**

From [auth0 documentation](https://auth0.com/docs/secure/security-guidance/data-security/user-data-storage#external-database-vs-auth0-data-store):

> For example, you could have a Users table that lists each user authenticated by Auth0. Every time a user logs in, you could search the table for that user. If the user does not exist, you would create a new record. If they do exist, you would update all fields, essentially keeping a local copy of all user data.

**Some more context**
- I deleted the migration that added membership types and just integrated it with the first migration.
- We now sync all user data from auth0 to the local table `ow_user` when a user authenticates (not exactly, it's actually in session callback).

User data from auth0 is synced with the respective row in the ow_user table **if it was last updated more than 24h ago** using the `session` callback. 

This number is somewhat arbitrary, so it should be reconsidered.

The`session`-callback is ran a lot. It runs on refocus (as the `useSession` is probably reran on window refocus). I think it is unneccesary to run the database insert that often.

Other callbacks we could use ([documentation](https://next-auth.js.org/configuration/callbacks#sign-in-callback)):
- `signin`
	- this might be invoked to rarely? not sure
- `redirect`
    - triggered as often as signIn more or less
-  `jwt`
	- From manual testing this is triggered a bunch of times on page load, so I don't think its any better than `session`

### CODE REVIEW

File: packages/auth/src/utils.ts
Comment (junlarsen) [inline packages/auth/src/utils.ts:1]: Can these go into a service? Auth0SynchronizationService or something

File: packages/core/src/lib/auth0-repository.ts
Comment (junlarsen) [inline packages/core/src/lib/auth0-repository.ts:10]:
```suggestion
export class Auth0RepositoryImpl implements Auth0Repository {
```

File: packages/core/src/lib/auth0-repository.ts
Comment (junlarsen) [inline packages/core/src/lib/auth0-repository.ts:43]:
```suggestion
  async getBySubject(sub: string): Promise<OidcUser | undefined> {
```

**Review by junlarsen — CHANGES_REQUESTED** [review]

File: packages/auth/src/utils.ts
Comment (henrikskog) [reply packages/auth/src/utils.ts:1]: Yes that makes more sense, good call! Where do you think this service belongs? With auth0-repository in `/core/src/lib`?

File: tools/migrator/src/fixtures/user.ts
Comment (henrikhorluck) [inline tools/migrator/src/fixtures/user.ts:26]: IMO these should not be real names/emails, and we should look into how to get a good devX here.

Also these sub's are for dev?

**Review by henrikhorluck — CHANGES_REQUESTED** [review]

File: packages/auth/src/utils.ts
Comment (henrikskog) [reply packages/auth/src/utils.ts:1]: I put it in `packages/core/src/lib/auth0-synchronization-service.ts` LMK what you think!

File: packages/auth/src/auth-options.ts
Comment (henrikskog) [reply packages/auth/src/auth-options.ts:45]: Nice! I guess some things are left behind from Cognito.

File: packages/auth/src/utils.ts
Comment (henrikskog) [reply packages/auth/src/utils.ts]: I agree. This is a somewhat lazy solution where I could not access the type in `packages/auth/src/auth-options.ts` due to it being in a module and I was not sure if I could change that or not. I'll take a closer look at this.

File: packages/db/src/migrations/0001_create_user_and_auth.js
Comment (henrikskog) [reply packages/db/src/migrations/0001_create_user_and_auth.js:14]: Good insights! I would like to discuss this more tomorrow, let's try not to forget

File: tools/migrator/src/fixtures/user.ts
Comment (henrikskog) [reply tools/migrator/src/fixtures/user.ts:26]: Yeah this setup is a bit awkward. My sub is for dev at least, I have not checked the others.

File: packages/core/src/modules/user/user-repository.ts
Comment (henrikskog) [reply packages/core/src/modules/user/user-repository.ts:40]: Good call!

File: packages/auth/src/utils.ts
Comment (junlarsen) [reply packages/auth/src/utils.ts:1]: Sure, I think that lib folder should probably be deleted at some point but that's ok for now

File: tools/migrator/src/fixtures/user.ts
Comment (henrikskog) [reply tools/migrator/src/fixtures/user.ts:28]: I created a new issue for this on Linear. I don't think it's something that should be fixed as a part of this PR.

File: packages/auth/src/utils.ts
Comment (henrikskog) [reply packages/auth/src/utils.ts]: I added an explanatory comment. We don't want core to depend on the auth package.

File: packages/core/src/lib/auth0-synchronization-service.ts
Comment (junlarsen) [inline packages/core/src/lib/auth0-synchronization-service.ts:26]: :nerd_face: :point_up: 
```suggestion
  createUser: (token: JWTToken) => Promise<User>

  /**
   * Synchronizes the provided user with data from Auth0. Only updates user if the user's updatedAt field is more than 1 day ago.
   */
  synchronizeUser: (user: User) => Promise<User | undefined>
```

**Review by junlarsen — APPROVED** [review]

File: packages/core/src/lib/auth0-synchronization-service.ts
Comment (henrikhorluck) [inline packages/core/src/lib/auth0-synchronization-service.ts:26]:
```suggestion
  synchronizeUser: (user: User) => Promise<User | null>
```

File: packages/core/src/lib/auth0-synchronization-service.ts
Comment (henrikhorluck) [inline packages/core/src/lib/auth0-synchronization-service.ts:74]: I would rather think you just perform a sync whenever you get an `id_token` (which is every time a user has to log in I think, and maybe whenever you refresh the access token? I do not know how much nextAuth does though), and that whenever the user wants to change something that we store in auth0, what is stored in auth0 is changed first, then we invalidate the `id_token` and fetch a new one? 

The 1 day is arbitrary and `id_token`s has it's own expiry (which by default is 10 hours). 

This kind of syncing just leads to a one-day gap between changing e.g. givenName and it actually being updated, if we forget to invalidate the id_token?

**Review by henrikhorluck — CHANGES_REQUESTED** [review]

File: packages/core/src/lib/auth0-repository.ts
Comment (henrikskog) [reply packages/core/src/lib/auth0-repository.ts:5]: True, thanks!

File: packages/auth/src/auth-options.ts
Comment (henrikskog) [reply packages/auth/src/auth-options.ts:84]: What do you think about this issue @junlarsen? In my opinion it is fine to use the sub as an id. Is there any arguments of having a normal ulid id on the table?

Comment (junlarsen) [reply packages/auth/src/auth-options.ts:84]: Either one works, I guess it might be annoying if we ever migrate off Auth0 but I'd say its fine to just use the Auth0 sub here

File: packages/core/src/lib/auth0-repository.ts
Comment (junlarsen) [inline packages/core/src/lib/auth0-repository.ts:19]: Create a zod scema in /packages/types and parse here

File: packages/core/src/lib/auth0-synchronization-service.ts
Comment (junlarsen) [inline packages/core/src/lib/auth0-synchronization-service.ts:57]: Remove this and replace with a logger from /packages/logger

**Review by junlarsen — CHANGES_REQUESTED** [review]

File: .env.example
Comment (henrikskog) [reply .env.example:11]: I can't find this in doppler though?

File: packages/auth/src/auth-options.ts
Comment (henrikskog) [reply packages/auth/src/auth-options.ts:84]: See https://linear.app/dotkom/issue/DOT-628/refactor-user-to-use-the-id-from-auth0, I think it's better to do this in a seperate PR

File: packages/core/src/lib/auth0-synchronization-service.ts
Comment (henrikskog) [reply packages/core/src/lib/auth0-synchronization-service.ts]: > I would rather think you just perform a sync whenever you get an id_token (which is every time a user has to log in I think, and maybe whenever you refresh the access token? I do not know how much nextAuth does though), and that whenever the user wants to change something that we store in auth0, what is stored in auth0 is changed first, then we invalidate the id_token and fetch a new one?

I created this issue for it: https://linear.app/dotkom/issue/DOT-629/find-a-better-solution-for-synchronizing-auth0-user-data-to-ow-user. I'm not quite sure how this is done and I feel blocked by this PR on my #804 PR. In my opinion this is a case of optimization in regards to the amount of database calls (we fetch the user from the database on every `session` callback). I don't see what synchronization problems it leads to. So I would rather just merge this now and find a more optimized solution later.

Thoughts?

> This kind of syncing just leads to a one-day gap between changing e.g. givenName and it actually being updated, if we forget to invalidate the id_token?

When updating givenName we will use the auth0 mgmt api, then sync the user with the database. We would also do it this way if we did not store any user info in the application database. In other words, we will have the same problem with invalidating tokens no matter what right? I don't see how this is relevant to this synchronization approach?

Comment (junlarsen) [reply packages/core/src/lib/auth0-synchronization-service.ts]: From what I can tell, a potential syncing problem would exist in either case?

I agree that it's probably easiest to merge this now, and worry about moving to synchronizing against id_token in a future PR

File: packages/core/logger.ts
Comment (junlarsen) [inline packages/core/logger.ts:3]: This is fine for now, but in the future we want a specific logger in each service where needed, that way it's a lot easier to trace

**Review by junlarsen — APPROVED** [review]
- [ ]

**Review by junlarsen — APPROVED** [review]
- [ ]

File: .env.example
Comment (henrikhorluck) [reply .env.example:11]: Then I will add it

Comment (henrikskog) [reply .env.example:11]: Okay nice

File: packages/core/src/lib/auth0-repository.ts
Comment (henrikhorluck) [inline packages/core/src/lib/auth0-repository.ts:10]: If we have an access token we can just use `/userinfo`, no need to use the mgmt API.

What is in the `id_token` may have a better schema than the mgmt-api (e.g. `gender` is actually gender instead of being in `user_metadata/gender`, and we _might_ add stuff to the id_token that is not in auth0

File: packages/db/src/migrations/0001_create_user_and_auth.js
Comment (henrikhorluck) [reply packages/db/src/migrations/0001_create_user_and_auth.js:15]: Changing old migrations is a bad habit to acquire, and if you feel like monoweb is making it hard to properly make migrations we should actually do something about it. 

This change means that everyone that develops on monoweb has to recreate their database-schema and delete all their data which I think is unnecessary, even though it does not cause any harm to any production db this is like throwing a stick into the wheel of other developers

> This way it is clear what the user table includes, rather than having it in many different migration files.

Having fewer migration files is not a solution, people should not be reading migrations to figure out the current db schema

**Review by henrikhorluck — CHANGES_REQUESTED** [review]

File: packages/core/src/lib/auth0-repository.ts
Comment (junlarsen) [reply packages/core/src/lib/auth0-repository.ts:10]: Probably for fetching attendees on an event?

Comment (henrikhorluck) [reply packages/core/src/lib/auth0-repository.ts:10]: If that user has not logged in, there is no way for the data to go stale?

File: packages/auth/src/auth-options.ts
Comment (henrikskog) [reply packages/auth/src/auth-options.ts]: Oops! There.

File: packages/core/src/lib/auth0-synchronization-service.ts
Comment (henrikskog) [reply packages/core/src/lib/auth0-synchronization-service.ts]: Mixed up my branches while revieweing, made the change on the other PR im working on. There it's fixed. Sorry about that.

File: packages/db/src/migrations/0001_create_user_and_auth.js
Comment (henrikskog) [reply packages/db/src/migrations/0001_create_user_and_auth.js:15]: I researched this more now and I realize that this change will indeed be very annoying for other people that are using their own database. 

Am I right in this thinking:

In this PR I am adding fields to the `ow_user` with the `not null` restriction. E.g. `name`.

The proper way of doing this would be to:
1. create new migration
2. create new fields WITHOUT not null restriction
3. modify all rows with values for the new fields that are not null
4. modify fields to have not null restriction

I have not been working with migrations before so for me this seems like a tideous process. However, maybe it's not, I don't know.

My workflow when creating these new migrations have been to just delete the database and create it before running migrations. But I understand this is not something people have set up.

File: packages/core/src/lib/auth0-repository.ts
Comment (henrikskog) [reply packages/core/src/lib/auth0-repository.ts:10]: What about this scenario:
1. User goes to their profile on our website and change their email
2. monoweb uses management api to update email in auth0

Now it is important that synchronizations after this actually use the new user state and not some stale data from id tokens. 

For the id tokens to contain the new user data, the user has to authorize again. Hitting the /userinfo endpoint will return an id token with stale data (old email before patch). Read [this](https://community.auth0.com/t/auth0-returns-a-stale-userinfo/13030/4)

If we just use the management api we can make sure we never use stale data from auth0 when synchronizing.

Thoughts?

Comment (henrikhorluck) [reply packages/core/src/lib/auth0-repository.ts:10]: Ah, that is kinda unexpected but also understandable. Does the id_token update if you refresh the access_token?

Comment (henrikskog) [reply packages/core/src/lib/auth0-repository.ts:10]: I don't know. I suggest we use the management api for now.

File: packages/db/src/migrations/0001_create_user_and_auth.js
Comment (henrikhorluck) [reply packages/db/src/migrations/0001_create_user_and_auth.js:15]: This is a pain to deal with if anything goes wrong, IMO devs _should just need to apply a migration_. It is valuable to have an examples of proper migration for later devs, and you benefit from learning how to do this properly. This kind of migration will be needed later as well.

I leave it to your discretion to find a suitable solution

File: packages/db/src/db.generated.d.ts
Comment (henrikhorluck) [inline packages/db/src/db.generated.d.ts:197]:
```suggestion
  lastSyncedAt: Timestamp
```
This should just be default now?

File: packages/db/src/migrations/0024_user.js
Comment (henrikhorluck) [inline packages/db/src/migrations/0024_user.js:16]:
```suggestion
    const email = `${user.id}@invalid.local`
```

**Review by henrikhorluck — APPROVED** [review]

File: packages/db/src/db.generated.d.ts
Comment (henrikskog) [reply packages/db/src/db.generated.d.ts:197]: Yea agree, good suggestion.

File: packages/db/src/migrations/0024_user.js
Comment (henrikskog) [reply packages/db/src/migrations/0024_user.js]: smart!

**Pull request merged**
